table_helper.cc

```diff
@@ -59,17 +59,31 @@ future<> table_helper::setup_table() const {
 }
 
 future<> table_helper::cache_table_info(service::query_state& qs) {
-    if (_prepared_stmt) {
-        return now();
-    } else {
+    if (!_prepared_stmt) {
         // if prepared statement has been invalidated - drop cached pointers
         _insert_stmt = nullptr;
+    } else if (!_is_fallback_stmt) {
+        return now();
     }
 
     return cql3::get_local_query_processor().prepare(_insert_cql, qs.get_client_state(), false).then([this] (shared_ptr<cql_transport::messages::result_message::prepared> msg_ptr) {
         _prepared_stmt = std::move(msg_ptr->get_prepared());
         shared_ptr<cql3::cql_statement> cql_stmt = _prepared_stmt->statement;
         _insert_stmt = dynamic_pointer_cast<cql3::statements::modification_statement>(cql_stmt);
+        _is_fallback_stmt = false;
+    }).handle_exception([this, &qs] (auto eptr) {
+        if (!_insert_cql_fallback) {
+            return make_exception_future(eptr);
+        }
+        if (_is_fallback_stmt && _prepared_stmt) {
+            return now();
+        }
+        return cql3::get_local_query_processor().prepare(_insert_cql_fallback.value(), qs.get_client_state(), false).then([this] (shared_ptr<cql_transport::messages::result_message::prepared> msg_ptr) {
+            _prepared_stmt = std::move(msg_ptr->get_prepared());
+            shared_ptr<cql3::cql_statement> cql_stmt = _prepared_stmt->statement;
+            _insert_stmt = dynamic_pointer_cast<cql3::statements::modification_statement>(cql_stmt);
+            _is_fallback_stmt = true;
```

Comment:
I have mixed feelings about this state. Is it safe when table_helper::insert (and then table_helper::cache_table_info) is called concurrently multiple times? Will it still behave correctly?

Reply:
Continuations on the same thread (so those who can access the same memory) do not run in parallel. Therefore each time a new value is assigned to _instert_stmt, the _is_fallback_stmt is changed accordingly at the same time.
I should probably add a comment that this is the intended use of the _is_fallback_stmt variable.

Reply:
I guess if we assume that none of the callbacks throws an exception then it should be ok.

Reply:
That's the assumption. Maybe I should have added the noexcept specifier to the lambdas.